src/qservkafka/storage/votable.py

```diff
@@ -236,6 +239,184 @@ def _encode_char_column(
+            millisecond = value_raw.microsecond // 1000
+            return (
+                f"{value_raw.year:04d}-{value_raw.month:02d}"
+                f"-{value_raw.day:02d}T{value_raw.hour:02d}"
+                f":{value_raw.minute:02d}:{value_raw.second:02d}"
+                f".{millisecond:03d}"
+            )
+        return "" if value_raw is None else str(value_raw)
+
+    def _apply_url_rewriting(
+        self, value_str: str, column: JobResultColumnType
+    ) -> str:
+        """Apply URL rewriting if configured.
+
+        Parameters
+        ----------
+        value_str
+            String value to potentially rewrite.
+        column
+            Column def which may indicate whether URL rewriting is needed.
+
+        Returns
+        -------
+        str
+            The rewritten URL or the original string.
+
+        """
+        if not (value_str and column.requires_url_rewrite):
+            return value_str
+
+        try:
+            base_url = urlparse(str(config.rewrite_base_url))
+            url = urlparse(value_str)
+            return url._replace(netloc=base_url.netloc).geturl()
+        except Exception as e:
+            self._logger.warning(
+                "Unable to rewrite URL", column=column.name, error=str(e)
+            )
+            return value_str
+
+    def _encode_unicode_null(self, column: JobResultColumnType) -> bytes:
+        """Encode null value for unicodeChar column.
+
+        Parameters
+        ----------
+        column
+            Column type definition.
+
+        Returns
+        -------
+        bytes
+            Serialized representation of a null value for unicodeChar.
+        """
+        if column.arraysize and column.arraysize.variable:
+            return _zero_int
+        elif column.arraysize and column.arraysize.limit:
+            max_bytes = column.arraysize.limit * 2
+            rule = str(max_bytes) + "s"
+            return struct.pack(rule, b"")
+        else:
+            return column.datatype.pack("")
+
+    def _encode_unicode_variable_array(
+        self, value_str: str, value: bytes, column: JobResultColumnType
+    ) -> bytes:
+        """Encode variable-length unicodeChar array.
+
+        Parameters
+        ----------
+        value_str
+            String value to encode.
+        value
+            Encoded byte representation of the string.
+        column
+            Column type definition.
+
+        Returns
+        -------
+        bytes
+            Serialized representation of the variable-length unicodeChar array.
+        """
+        if (
+            column.arraysize
+            and column.arraysize.limit
+            and len(value_str) > column.arraysize.limit
+        ):
+            value_str = value_str[: column.arraysize.limit]
+            value = value_str.encode("utf-16-be")
+
+        char_count = len(value) // 2
+        rule = ">I" + str(len(value)) + "s"
+        return struct.pack(rule, char_count, value)
```

Comment:
This encoding logic will generate a column longer than `limit` in cases where the UTF-16 encoding involves surrogate pairs. I think the truncation really does have to be done on the encoded value, and thus there's no need to pass `value_str` into this function.

Reply:
To expand on what I said above, I think the client *can't* treat arraysize as Unicode characters because that would require doing UTF-16 decoding while doing BINARY2 decoding and pull out octets for the `unicodeChar` string based on the partial decoding after seeing if it's the first half of a surrogate pair. I'm pretty sure no client will do that since the standard says the output is in UCS-2.

This I think is part of why Marcus was arguing for just deprecating `unicodeChar` as a lost cause.

And yes, if you truncate after encoding, you have to handle the case where the last character is a surrogate pair and you would split that pair in half, and instead truncate before the surrogate pair. It's a pain in the ass, which is why UTF-16 is annoying when the standard says it's UCS-2.

Reply:
Ok, I've updated the two encoding methods to do byte-level truncation and added a helper method to truncate and detect/remove orphaned surrogates. I've added some checks for this, but perhaps you can double check that I'm doing the right thing,